### PR TITLE
Expose `determinant` in Transform2D, rename internal method

### PR DESCRIPTION
**Pull request by aaronfranke**
I usually use 3D. As a math nerd I was helping out a friend with some math in a 2D game and I was shocked that the determinant was not exposed to GDScript. This PR exposes it.

Discussion: Since this is not exposed yet, we could also rename it from `basis_determinant` to just `determinant`, since there is no other possible way to interpret the word "determinant" here. If this is desired I would be happy to add it to this PR, but I don't want to cause unnecessary friction so I haven't done so yet. `basis_determinant` is still clear and correct, it's just... kinda redundant and unnecessary.

3.x version: https://github.com/godotengine/godot/pull/76323

### CODE REVIEW

**Comment by MewPurPur** [comment]
I agree both with exposing and renaming it. It doesn't have anything to do with the Basis class, does it?

**Comment by kleonc** [comment]
> Discussion: Since this is not exposed yet, we could also rename it from `basis_determinant` to just `determinant`, since there is no other possible way to interpret the word "determinant" here. If this is desired I would be happy to add it to this PR, but I don't want to cause unnecessary friction so I haven't done so yet. `basis_determinant` is still clear and correct, it's just... kinda redundant and unnecessary.

Well, determinant makes sense only for square matrices. And [Transform2D is said to be a 2×3 matrix](https://github.com/godotengine/godot/blob/6f1a52b017930afb851691e6852ce40eb93c30a9/doc/classes/Transform2D.xml#L4-L7). Given that, `Transform2D.determinant` seems kinda incorrect. It could be interpreted as either determinant of the basis (the 2×2 part of the matrix, the first two columns), or as a determinant of the 3x3 matrix the given Transform2D represents (for which the third non-stored row is always assumed to be `[0, 0, 1]`). But for both of the interpretations the result is the same, so there's no discrepancy on that part.
I find `Transform2D.basis_determinant` would be more conceptually correct. But `Transform2D.determinant` is shorter and should be fine too, given the docs will clearly explain what it actually returns.
Conceptually e.g. `Vector2.cross` also makes no sense but we do have it. :upside_down_face: So yeah, `determinant` + explanation in the docs should be fine.

> I agree both with exposing and renaming it. It doesn't have anything to do with the Basis class, does it?

@MewPurPur It does not, it has to do with [the basis part](https://en.wikipedia.org/wiki/Basis_(linear_algebra)) of the Transform2D.

**Comment by MewPurPur** [comment]
Hmm, sure thing, but the determinant tells us what the area of a unit square would be under the matrix transformation and this is unaffected when we also have an origin. For xform it's more relevant to tell this than for determinant.

**Comment by aaronfranke** [comment]
Pushed an update to rename to `determinant`. One more argument in favor of the rename:

* We also have `orthonormalized`, `get_rotation`, `get_scale`, `get_skew`, `rotated_local`, `scaled_local`, which only use the basis don't include "basis" in the name. The only other methods that do include it are the xform methods which are necessary to disambiguate it from a full transform using the origin.

@MewPurPur Great point, it makes perfect sense conceptually when thinking about how the area changes, which is not affected by a translation.

> Conceptually e.g. `Vector2.cross` also makes no sense but we do have it. 🙃

[It does make sense actually](https://www.youtube.com/watch?v=Ip3X9LOh2dk) in fact [you can even extend it into 4D](https://github.com/aaronfranke/godot/tree/4d-cross).

**Comment by MewPurPur** [comment]
I'd consider Vector2.cross a shorthand notation for a mathematical hack, same for all cross products that aren't 3D or 7D

**Comment by kleonc** [comment]
> Hmm, sure thing, but the determinant tells us what the area of a unit square would be under the matrix transformation and this is unaffected from changing the origin. For xform it's more relevant to tell this than for determinant.

@MewPurPur What I was saying is that without a square matrix there's no talk about the determinant in the first place. Determinant is defined only for square matrices. By saying "determinant" you're automatically referring to some square matrix (the 2x2 or 3x3 matrix I've mentioned in https://github.com/godotengine/godot/pull/76311#issuecomment-1517683783).

---

> * We also have `orthonormalized`, `get_rotation`, `get_scale`, `get_skew`, `rotated_local`, `scaled_local`, which only use the basis don't include "basis" in the name. The only other methods that do include it are the xform methods which are necessary to disambiguate it from a full transform using the origin.

@aaronfranke Yeah, I've realized this about `get_scale` too after writing my previous comment, didn't bother with editing my comment as I've already agreed that `determinant` is fine anyway.

> > Conceptually e.g. `Vector2.cross` also makes no sense but we do have it. upside_down_face
> 
> [It does make sense actually](https://www.youtube.com/watch?v=Ip3X9LOh2dk)

@aaronfranke Was its supposed to link to a video about determinant? I'd expect something `Vector2.cross` related based on the quote. :upside_down_face: Anyway, if the link is correct then note that in the video he's referring only to linear transformations / square matrices.

---

> I'd consider Vector2.cross a shorthand notation for a mathematical hack, same for all cross products that aren't 3D or 7D. ... But that's off-topic :P

The 2D case can be seen as a simplified 3D case with vectors' third component being zero, so it could be said to be a simplification and not really a hack though. Unless that's exactly what you call "a mathematical hack". :smile:
(We're indeed going offtopic and I've already agreed `Transform2D.determinant` is fine too, so I'll cut it here :slightly_smiling_face:)

File: doc/classes/Transform2D.xml [hunk 1]
```diff
@@ -78,6 +78,13 @@
 				This method does not account for translation (the origin vector).
 			</description>
 		</method>
+		<method name="determinant" qualifiers="const">
+			<return type="float" />
+			<description>
+				Returns the determinant of the basis matrix. If the basis is uniformly scaled, its determinant is the square of the scale.
```
Comment (kleonc) [inline doc/classes/Transform2D.xml:84]: Not sure if the second sentence is a vital information. :thinking: But if anything I'd reword it something like:
```suggestion
				Returns the determinant of the basis matrix. If the basis is uniformly scaled, then its determinant equals the square of the scale factor.
```
- "then" I think it's tend to be used explicitely within the docs, seen it many times like that. Not sure about the change.
- "equals" seems more fitting to me.
- "scale factor" `scale` is still kinda a two component vector, I find "scale factor" clearer (it refers to the scalar).

File: doc/classes/Transform2D.xml [hunk 2]
```diff
@@ -78,6 +78,13 @@
...
+				A negative determinant means the basis has a negative scale. A zero determinant means the basis isn't invertible, and is usually considered invalid.
```
Comment (kleonc) [inline doc/classes/Transform2D.xml:85]: The first sentence: what does "a negative scale" is meant to mean? Because for scale like `(-1, -1)` the determinant would be positive. Negative determinant means the scale's components has opposite signs (and both are non-zero).

**Comment by aaronfranke** [comment]
@kleonc Oh, sorry, I linked the wrong video. I meant to link [this video](https://www.youtube.com/watch?v=eu6i7WJeinw). But yeah we already agree :)

File: doc/classes/Transform2D.xml (hunk 2 above)
Comment (aaronfranke) [reply doc/classes/Transform2D.xml:85]: Technically a scale of (-1, -1) can't be decomposed because it would be interpreted as a 180 degree rotation instead. However, indeed we can be more clear here, I'll update the text.

**Review by kleonc — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
